hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/checksum/TestContainerChecksumTreeManager.java

```diff
@@ -293,12 +319,170 @@ public void testEmptyFile() throws Exception {
+        .addAllDeletedBlocks(deletedBlockList);
+
+    ContainerProtos.ContainerChecksumInfo peerChecksumInfo = peerChecksumInfoBuilder.build();
+    checksumManager.writeContainerDataTree(container, ourMerkleTree);
+    ContainerChecksumTreeManager.ContainerDiff containerDiff = checksumManager.diff(container, peerChecksumInfo);
+
+    // The diff should not have any missing block/missing chunk/corrupt chunks as the blocks are deleted
+    // in peer merkle tree.
+    assertTrue(containerDiff.getMissingBlocks().isEmpty());
+    assertTrue(containerDiff.getMissingChunks().isEmpty());
+    assertTrue(containerDiff.getMissingChunks().isEmpty());
+
+    // Delete blocks in our merkle tree as well.
+    checksumManager.markBlocksAsDeleted(container, deletedBlockList);
+    containerDiff = checksumManager.diff(container, peerChecksumInfo);
+
+    // The diff should not have any missing block/missing chunk/corrupt chunks as the blocks are deleted
+    // in both merkle tree.
+    assertTrue(containerDiff.getMissingBlocks().isEmpty());
+    assertTrue(containerDiff.getMissingChunks().isEmpty());
+    assertTrue(containerDiff.getMissingChunks().isEmpty());
+  }
+
+  /**
+   * Test to check if the container diff consists of blocks that are corrupted in our merkle tree but also deleted in
+   * our merkle tree.
+   */
+  @Test
+  void testDeletedBlocksInOurContainerOnly() throws Exception {
+    // Setup deleted blocks only in the peer container checksum
+    ContainerMerkleTree peerMerkleTree = buildTestTree(config);
+    // Introduce block corruption in our merkle tree.
+    ContainerProtos.ContainerMerkleTree ourMerkleTree = buildTestTreeWithMismatches(peerMerkleTree, 0, 3, 3).getLeft();
+    List<Long> deletedBlockList = Arrays.asList(1L, 2L, 3L, 4L, 5L);
+    ContainerProtos.ContainerChecksumInfo.Builder peerChecksumInfoBuilder = ContainerProtos.ContainerChecksumInfo
+        .newBuilder().setContainerMerkleTree(ourMerkleTree).setContainerID(CONTAINER_ID);
+
+    ContainerProtos.ContainerChecksumInfo peerChecksumInfo = peerChecksumInfoBuilder.build();
+    checksumManager.writeContainerDataTree(container, ourMerkleTree);
+    checksumManager.markBlocksAsDeleted(container, deletedBlockList);
+
+    ContainerChecksumTreeManager.ContainerDiff containerDiff = checksumManager.diff(container, peerChecksumInfo);
+
+    // The diff should not have any missing block/missing chunk/corrupt chunks as the blocks are deleted
+    // in our merkle tree.
+    assertTrue(containerDiff.getMissingBlocks().isEmpty());
+    assertTrue(containerDiff.getMissingChunks().isEmpty());
+    assertTrue(containerDiff.getMissingChunks().isEmpty());
+  }
+
+  /**
+   * Test to check if the container diff consists of blocks that are corrupted in our merkle tree but also deleted in
+   * our peer tree.
+   */
+  @Test
+  void testCorruptionInOurMerkleTreeAndDeletedBlocksInPeer() throws Exception {
```

Comment:
These deleted block tests are good, but can we add another that tests when only some blocks are marked as deleted that the remaining blocks still get flagged correctly?